### PR TITLE
fast static require

### PR DESCRIPTION
**Pull request by grosser**
@indirect got inspired by @tenderlove's red-dot talk to make require O(1) ... took our app boot time from 18s->12s (worked for ~3600 of ~4000 requires = 90% cache hits with 14k files in the cache total) ... are you interested in adding something like this ... obviously needs more work / tests, but wanted to get your feedback

### CODE REVIEW

**Comment by indirect** [comment]
Looks like #3882 is relevant to your interests. I'm not yet sure about the tradeoffs between performance and correctness in these options; if we can make require more performant without any loss of correctness, that would be pretty fantastic.

**Comment by byroot** [comment]
Well in #3882 I took it a bit farther by wanting not to populate the $LOAD_PATH. But a tamer version that would let the LOAD_PATH would still provide a big speedup while being 100% correct.

The downside is that the correctness is tied on reimplementing the require precedence logic in ruby. I mean this: https://github.com/byroot/bundler/blob/3fc5230b7956a148558b545d7c12c50040302683/lib/bundler/require_cache.rb#L42-L61

**Comment by segiddins** [comment]
I think this approach is cool, but seeing as it has absolutely no bundler-specific code... I'm just unsure it belongs in Bundler.

**Comment by grosser** [comment]
In a non-bundler app this simple approach makes less sense since the load-path is added to all the time, also it has to be done after bundler/setup ... so I figured here is a good place ... maybe only to kick off a discussion / being told 'this already exists' :)

@indirect this loses correctness once something is added/removed to the LOAD_PATH ... but we could have a 'strict' mode where it blows up when the load-path got changed or warn or 'stop being fast' to be correct.

I'm not worried about native gems etc foobar since they only make up a small percentage of requires -> lets get that easy 90% :)

**Comment by grosser** [comment]
added verification, so you can optionally make sure it is correct

**Comment by grosser** [comment]
so if someone is worried about correctness they would set it to raise, boot their app and fix all the extra load-paths that get added in random places

```
require bundler/setup'
$LOAD_PATH << 'custom_foo'
require bundler/static_setup'
```

File: lib/bundler/static_setup.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
+require 'bundler/setup'
+
+# cache rack -> /foo/bar/vendor/bundle/rack/lib/rack.rb
+module Bundler
+  PRELOADED_REQUIRE = {}
+  $LOAD_PATH.reverse_each do |path|
+    Dir["#{path}/**/*.rb"].each do |f|
+      start = path.size + 1
+      count = f.size - start - 3 # without '.rb' = last 3
```
Comment (simi) [inline lib/bundler/static_setup.rb]: Is this (`.rb`) really universal?

Comment (grosser) [inline lib/bundler/static_setup.rb]: with just .rb hits 4698 - misses 405
with .so,.o added it's the same ... so seems to not make any difference ...
and even if it does probably not worth the extra complexity/time (search
with pattern + extracting extname) ... on miss it falls back to default
ruby anyway

**Comment by byroot** [comment]
Since it seems complicated to have a patch like this in bundler or rubygems, I published my version of this trick as a gem: https://github.com/byroot/bootscale. I think it's simple enough to integrate for people who want to do so.

@grosser if you'd like to try to integrate that gem in Zendesk, I'd love to hear your feedback. I tested it on Shopify in replacement of our old patch and it works fine.

**Comment by grosser** [comment]
The static Kernel patch did not help me, but the autoload patch fixed
another 300 files :)

Seems to work nicely! Thx for gemifying :)

**Comment by byroot** [comment]
> The static Kernel patch did not help me

It does pick up a relatively low number of files for me. I'll try to figure out what the difference is.

> Seems to work nicely! Thx for gemifying :)

My pleasure!

**Pull request closed without merging**